src/contracts/utils/WrapAndNiftyswap.sol

```diff
@@ -1,163 +1,149 @@
+        exchange = _exchange;
+        erc20 = _erc20;
+        erc1155 = _erc1155;
+
+        // Approve wrapper contract for ERC-20
+        // NOTE: This could potentially fail in some extreme usage as it's only
+        // set once, but can easily redeploy this contract if that's the case.
+        IERC20(_erc20).approve(_tokenWrapper, 2 ** 256 - 1);
+
+        // Store wrapped token ID
+        wrappedTokenID = IERC20Wrapper(_tokenWrapper).getTokenID(_erc20);
     }
 
-    // Transfer tokens purchased
-    IERC1155(erc1155).safeBatchTransferFrom(address(this), _recipient, obj.tokensBoughtIDs, obj.tokensBoughtAmounts, "");
-  }
-
-  /**
-   * @notice Accepts only tokenWrapper tokens 
-   * @return `bytes4(keccak256("onERC1155BatchReceived(address,address,uint256[],uint256[],bytes)"))`
-   */
-  function onERC1155Received(address, address, uint256, uint256, bytes calldata)
-    external returns(bytes4)
-  {
-    if (msg.sender != address(tokenWrapper)) {
-      revert("WrapAndNiftyswap#onERC1155Received: INVALID_ERC1155_RECEIVED");
-    }
-    return IERC1155TokenReceiver.onERC1155Received.selector;
-  }
-
-  /**
-   * @notice If receives tracked ERC-1155, it will send a sell order to niftyswap and unwrap received
-   *         wrapped token. The unwrapped tokens will be sent to the sender.
-   * @return `bytes4(keccak256("onERC1155BatchReceived(address,address,uint256[],uint256[],bytes)"))`
-   */
-  function onERC1155BatchReceived(
-    address, 
-    address _from, 
-    uint256[] calldata _ids, 
-    uint256[] calldata _amounts, 
-    bytes calldata _niftyswapOrder
-  )
-    external returns(bytes4)
-  { 
-    // If coming from niftyswap or wrapped token, ignore
-    if (isInNiftyswap || msg.sender == address(tokenWrapper)){
-      return IERC1155TokenReceiver.onERC1155BatchReceived.selector;
-    } else if (msg.sender != erc1155) {
-      revert("WrapAndNiftyswap#onERC1155BatchReceived: INVALID_ERC1155_RECEIVED");
+    /**
+     * @notice Wrap ERC-20 to ERC-1155 and swap them
+     * @dev User must approve this contract for ERC-20 first
+     * @param _maxAmount       Maximum amount of ERC-20 user wants to spend
+     * @param _recipient       Address where to send tokens
+     * @param _niftyswapOrder  Encoded Niftyswap order passed in data field of safeTransferFrom()
+     */
+    function wrapAndSwap(uint256 _maxAmount, address _recipient, bytes calldata _niftyswapOrder) external {
+        // Decode niftyswap order
+        INiftyswapExchange.BuyTokensObj memory obj;
+        (, obj) = abi.decode(_niftyswapOrder, (bytes4, INiftyswapExchange.BuyTokensObj));
+
+        // Force the recipient to not be set, otherwise wrapped token refunded will be
+        // sent to the user and we won't be able to unwrap it.
+        require(
+            obj.recipient == address(0x0) || obj.recipient == address(this),
+            "WrapAndNiftyswap#wrapAndSwap: ORDER RECIPIENT MUST BE THIS CONTRACT"
```

Comment:
maybe we can take this chance to rename all revert messages similar to 

https://github.com/0xsequence/niftyswap/blob/be69ffd8576bb25b1a234b13eacdddeff1114d4a/src/contracts/exchange/NiftyswapExchange20.sol#L146

Reply:
was originally done for contract code size limit

Reply:
Ok, I'll only update to `NE#XX` for now